examples/models/animation.lisp

```diff
@@ -50,10 +49,18 @@
           (update-camera camera)
           (when (is-key-down-p +key-space+)
             (incf anim-frame-counter)
-            (update-model-animation model anim-frame-counter)
+            (update-model-animation model 0 anim-frame-counter)
             (when (>= anim-frame-counter (frame-count (elt (animations model) 0)))
               (setf anim-frame-counter 0)))
           (with-drawing ()
             (with-3d-mode camera
+              (loop for i below (length (bones model))
+                    ;; Sometimes it makes sense to call on low-level claylib (claylib/ll)
+                    do (claylib/ll:draw-cube
+                        (trans (elt (elt (frame-poses (elt (animations model) 0))
+                                         anim-frame-counter)
+                                    i))
+                        0.2 0.2 0.2
+                        (claylib::c-struct +red+)))
```

Comment:
Is this related to our instancing discussion or is there some other reason you chose to do it this way?

Regardless, this at the very least should be a pass-through. But if looping over a bones list this way is going to be a common pattern, I wonder if we could wrap that up somehow.

Reply:
Yes this is the source of that discussion. However, when we talked about it I didn't bring up the complexity of continuously referring to data that belongs to other objects. Here we are drawing a bunch of cubes but their positions are entirely in-sync with the translation of each bone. With the instance stuff we discussed we could share the size (0.2 0.2 0.2) of a reference cube among the instances, but would we have to make new vectors for the positions just to update them to already existing vectors? I could easily be wrong but `claylib/ll` seems like the place to go when you need to just tell a an "object" (like a cube, not an actual lisp object in this case) to look at an existing place in memory for some data.

I'm not sure if looping over bones will be a common pattern, but, if needed, that is one of the good things about having the rl-sequences. We can always implement [iterators](https://shinmera.github.io/trivial-extensible-sequences/#GENERIC-FUNCTION%20ORG.SHIRAKUMO.TRIVIAL-EXTENSIBLE-SEQUENCES%3AMAKE-SEQUENCE-ITERATOR) for the rl-sequences when we find the need.

Reply:
Oh I may have misunderstood you. You wrote looping over the bones *in this way*, meaning specifically accessing the translations of each bone in a given frame? I'd guess that's very rare, but I also think the nested `elt` calls are ugly. I was wondering if there was a way to do something akin to aref but with arbitrary combinations of arrays within arrays but I dunno if that would be better/cleaner/easy/possible.

Reply:
Any time we're doing skeletal animation, we'll be looping over bones and meshes (either we will, or the users will). I was mostly referring to `draw-cube`, which looked really weird to me at first glance, until I realized this example is basically using that as a substitute for normal meshes. Basically, I think we should wrap this up in a way such that the user only needs to pass in a frame counter and an animation number. e.g.: `(draw-pose my-model 0 42)`

The `draw-cube` bit is still weird, though. Does this model have actual drawable meshes associated? That would be a typical usage and I wonder if we should really be including such an unusual (IMHO) case in this example.

Reply:
Sorry. My previous comment was completely wrong and you should ignore it. Let's try again...

So here we have a case where a small-ish number of extra cubes is being drawn, but their positions are tied to the model's frame poses. Approaches:

1. Basically what you did, except make `draw-cube` a pass-through. Pros: Straightforward, non-contrived. Cons: Loss of some interactivity/inspectability at the REPL.
2. Create ~8 cube objects, set their position vectors `EQ` to the transforms each frame. Pros: Better interactivity/inspectability. Cons: Somewhat more contrived, possibly more consing.
3. Create ~8 cube objects, set their position _vector values_ to the transform values each frame. Pros: No extra consing. Cons: Way more contrived.
4. Implement instancing on the CLOS side. Pros: Could solve all of the above issues with none of the drawbacks. Cons: Will take time and I'm not 100% sure what it will look like yet.

Other possibilities?

Fundamentally -- now that we've redone memory management and scenes basically just control when assets are loaded and when GC is done, I don't have too much of a problem exporting functions like `draw-cube` if people choose to use them. That said, it's also not really "the Claylib way" and I would be hesitant to do that in our official examples unless there's really no better alternative. However, I am open to the possibility that "the Claylib way" itself needs to be tweaked.

Reply:
I think we should probably have instancing in the long term. But I thought this specific example is a good place to showcase the use of `claylib/ll`. It also may not be a good application of instancing since we still have to set position vectors (either like solution 2. or 3. above) for each different instance instead of just reading the values as `claylib/ll` does.

But yeah, the instancing in some form similar to what you had proposed earlier seems like a good idea at some point.

Reply:
Okay, that's fair. My only teensy quibble is that the last arg should be `claylib/ll:+red+` instead of grabbing the c-struct.

And I'll state for the record that users should be calling `claylib/ll` because they prefer it, not because `claylib` doesn't offer a better approach. In this case we've exposed a bit of a design issue that we'll have to iron out later, even if all that means is we create a `draw-cube` passthrough.

So fix that one line and let me know about the futures issue, then this will be ready to merge!

Reply:
> And I'll state for the record that users should be calling claylib/ll because they prefer it, not because claylib doesn't offer a better approach. In this case we've exposed a bit of a design issue that we'll have to iron out later, even if all that means is we create a draw-cube passthrough.

Okay, then I've [just switched](https://github.com/defun-games/claylib/pull/58/commits/7d3d490115f8bda187b25b3e6f6fcd05775ea855) this to a PT and I guess we'll want to add PTs for all `draw-*` functions so claylib doesn't lack claylib/ll functionality.

As for futures, I merged your `make-scene` changes but I am getting an issue where, in an object definition, I'm trying to get `(asset some-asset)` but it's returning `nil`, implying the asset has yet to be loaded. Looking into it further now.

Reply:
> As for futures, I merged your `make-scene` changes but I am getting an issue where, in an object definition, I'm trying to get `(asset some-asset)` but it's returning `nil`, implying the asset is not being loaded in time (or even that an unloaded version of it is brought into scope?). Looking into it further now.

With the animation example? That one is working for me, and in fact, all the examples are working for me, as I tested them before merging the `make-scene` changes. Assets are always loaded before objects and params are yielded so I'd be surprised if it's an ordering issue.

Reply:
Yeah, the animation example is working fine. I was having an issue with the cubicmap example - but switching to `make-scene-pro` (which I needed to do anyway for the camera) fixed it for some reason.

Should be ready to merge now :rocket: